File: modules/client/src/main/java/com/eventhorizon/pulsar/client/PulsarClient.java
package com.eventhorizon.pulsar.client;

import com.eventhorizon.pulsar.client.context.ServerContext;
import com.eventhorizon.pulsar.client.service.impl.BlueprintService;
import com.eventhorizon.pulsar.client.service.impl.ServerService;
import com.eventhorizon.pulsar.client.packet.PacketDispatcher;
import com.eventhorizon.pulsar.core.server.Server;
import com.eventhorizon.pulsar.core.server.ServerStatus;
import lombok.Getter;
import okhttp3.OkHttpClient;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tools.jackson.databind.json.JsonMapper;

import java.io.IOException;
import java.time.Duration;

@Getter
public final class PulsarClient {

    private static final Logger LOGGER = LoggerFactory.getLogger(PulsarClient.class);

    @Getter
    private static PulsarClient instance;

    private final BlueprintService blueprintApiClient;
    private final ServerService serverService;
    private final PacketDispatcher packetDispatcher;

    public PulsarClient(
            @NotNull String baseUrl,
            @NotNull String redisURI
    ) {
        this(baseUrl, redisURI, false);
    }

    public PulsarClient(
            @NotNull String baseUrl,
            @NotNull String redisURI,
            boolean debugMode
    ) {
        if (instance != null) {
            throw new IllegalStateException("PulsarClient is already initialized.");
        }

        final var okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(Duration.ofSeconds(5))
                .readTimeout(Duration.ofSeconds(30))
                .writeTimeout(Duration.ofSeconds(30))
                .retryOnConnectionFailure(true)
                .build();

        final var objectMapper = JsonMapper.builder()
                .findAndAddModules()
                .build();

        final var singularityAccessToken = System.getenv("SINGULARITY_ACCESS_TOKEN");

        this.blueprintApiClient = new BlueprintService(baseUrl, singularityAccessToken, okHttpClient, objectMapper);
        this.serverService = new ServerService(baseUrl, singularityAccessToken, okHttpClient, objectMapper);
        this.packetDispatcher = new PacketDispatcher(redisURI, debugMode);

        instance = this;
    }

    public boolean start() {
        final var serverId = System.getenv("SERVER_ID") != null ?
                System.getenv("SERVER_ID") :
                System.getProperty("SERVER_ID");

        if (serverId == null) {
            throw new IllegalArgumentException("SERVER_ID is not set. This process must be started by Singularity.");
        }

        try {
            final var server = this.serverService.get(serverId);
            if (server == null) {
                throw new IllegalArgumentException("SERVER_ID is invalid. This process must be started by Singularity.");
            }

            ServerContext.setServer(server);

            this.serverService.updateStatus(ServerContext.getDiscriminator(), ServerStatus.ACTIVE);
            this.packetDispatcher.start();
            return true;
        } catch (IOException e) {
            LOGGER.error("An unexpected error occurred, stopping Pulsar Client.", e);
            return false;
        }
    }

    public void stop() {
        this.packetDispatcher.stop();
    }
}

File: modules/client/src/main/kotlin/com/eventhorizon/pulsar/client/PulsarClient.kt
package com.eventhorizon.pulsar.client

import com.eventhorizon.pulsar.client.api.BlueprintApiClient
import com.eventhorizon.pulsar.client.api.ServerApiClient
import com.eventhorizon.pulsar.client.context.ServerContext
import com.eventhorizon.pulsar.client.packet.PacketClient
import com.eventhorizon.pulsar.client.packet.processor.scope.PulsarNetworkScope
import com.eventhorizon.pulsar.core.server.Server
import io.ktor.client.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json

class PulsarClient(
    endpointUrl: String,
    secretAccessToken: String,

    private val redisURI: String,
    private val maximumPacketSimultaneously: Int = 32,
    private val debugMode: Boolean = false
) {
    internal val httpClient = HttpClient(OkHttp) {
        install(Logging)
        install(ContentNegotiation) {
            json(
                Json {
                    ignoreUnknownKeys = true
                    isLenient = true
                }
            )
        }
    }

    val blueprints = BlueprintApiClient(httpClient, endpointUrl, secretAccessToken)
    val servers = ServerApiClient(httpClient, endpointUrl, secretAccessToken)

    lateinit var packets: PacketClient
        private set

    fun start() {
        val serverId = requireServerId()
        val server = runBlocking { requireServer(serverId) }

        ServerContext.server = server

        packets = PacketClient(
            redisURI,
            maximumPacketSimultaneously,
            debugMode
        )

        packets.start()
    }


    fun stop() {
        PulsarNetworkScope.launch {
            servers.delete(ServerContext.discriminator)
        }

        PulsarNetworkScope.cancel("Client shutdown.")
        httpClient.close()
    }

    private fun requireServerId(): String {
        return System.getenv("SERVER_ID")
            ?: System.getProperty("SERVER_ID")
            ?: error("SERVER_ID is not defined. This process must be started by Singularity.")
    }

    private suspend fun requireServer(serverId: String): Server =
        servers.get(serverId)
            ?: error("Server '$serverId' not found in Singularity.")

}